### PR TITLE
Feat: IpfsApiAccessError

### PR DESCRIPTION
**Pull request by JonKrone**
Use a common `Error` type, `IpfsApiAccessError`, when an app is denied access to a window.ipfs api.

Discussed in https://github.com/ipfs-shipyard/ipfs-companion/issues/455.

Remaining:
- [ ] Make this type available somewhere under `window.ipfs`. Perhaps the `ipfs.types` or `ipfs.utils` 
 from https://github.com/ipfs-shipyard/ipfs-companion/issues/444?

This also fixes some `path.sep` related test failures on Windows, lemme know what you think about that.

### CODE REVIEW

**Comment by JonKrone** [comment]
> Make this type available somewhere under window.ipfs. Perhaps the `ipfs.types` or `ipfs.utils`
from #444?

Looking for thoughts here.

**Comment by lidel** [comment]
### Browserify issue?

While it seems to work in tests, real life version injected as content script does not seem to produce Error with additional attributes. Try executing this against some website (and deny access to trigger error): 

```js
ipfs.files.add(Buffer.from('seed')).catch(e => console.log(`Error.permission=${e.permission}, scope=${e.scope}, message=${e.message}`))
```
I get:

screenshot_4

Are you able to reproduce?

### Namespace for proxy stuff

The `ipfs.types` sounds like an obvious place, but I am worried about unexpected consequences of polluting original API namespaces with proxy-specific stuff.

Perhaps a safer alternative is to use `window.ipfs.proxy` (idea suggested in https://github.com/ipfs-shipyard/ipfs-companion/issues/454#issuecomment-381278859) and expose it at `ipfs.proxy.types.IpfsApiAccessError`.

@olizilla @alanshaw do we have any strong opinions or preferred conventions for handling this?

**Comment by alanshaw** [comment]
Yes, I posted my comments here https://github.com/ipfs-shipyard/ipfs-companion/issues/455#issuecomment-381562091

**Comment by alanshaw** [comment]
On the posix stuff, I did check this originally because actually it needs to be posix.

The path module as included by browserify is https://github.com/substack/path-browserify/blob/master/index.js which seems to be using the posix versions and actually doesn't include a `path.posix`.

**Comment by lidel** [comment]
@JonKrone to limit the number of moving pieces, let's go with version proposed by @alanshaw  in https://github.com/ipfs-shipyard/ipfs-companion/issues/455#issuecomment-381562091 (no new type, just a self-explanatory property):
```js
if (err.isIpfsProxyAclError) {
```

That way we don't have to change anything on the backend and we don't need to think where to put the new type.

**Comment by JonKrone** [comment]
Hey guys, I'm sorry for the delay! I was totally swamped last week.

I pushed some changes a few days ago, thanks @alanshaw for the solution with ipfs-postmsg.

re: the posix path fix for tests on Windows: I explicitly imported `path-browserify`. We could alternatively manually normalize the paths, let me know what you think.

File: add-on/src/lib/ipfs-proxy/pre-mfs-scope.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
-const Path = require('path')
+// Use path-browserify for consistent behavior between browser and tests on Windows
```
Comment (lidel) [inline add-on/src/lib/ipfs-proxy/pre-mfs-scope.js]: I think explicitly using `path-browserify` (which provides posix version by default) is bit safer than assuming browserify-ing process will swap the implementation. 

Let's merge this, and if at some point `path-browserify` turns out to be not enough, we can swap it with own implementation / wrapper.

@alanshaw  does it sound ok to you?

File: docs/window.ipfs.md [hunk 2]
```diff
@@ -38,19 +38,19 @@ if (window.ipfs) {
...
-
-   const data = await ipfs.cat('QmS5RzFV9v4fucVtjATPxoxABgEmNhhduykzUbQeGyyS3N')
-   console.log(data.toString()) // =^.^=
+    const [{ hash }] = await ipfs.add(Buffer.from('=^.^='))
+    const data = await ipfs.cat(hash)
+    console.log(data.toString()) // =^.^=
+  } catch (err) {
+    if (err.isIpfsProxyAclError) {
+      // Fallback
+      console.log(':(')
```
Comment (lidel) [inline docs/window.ipfs.md]: I would replace `console.log(':(')` with `console.log('Unable to get ACL decision from user :(', err)`

**Pull request merged**